include/quicr/message_buffer.h

```diff
@@ -106,87 +104,145 @@ class MessageBuffer
+  std::vector<uint8_t> front(uint16_t len) const;
   std::vector<uint8_t> pop_front(uint16_t len);
 
+  /**
+   * @brief Moves the whole buffer, leaving MessageBuffer empty.
+   * @returns The buffer as an rvalue-ref.
+   */
   [[deprecated("quicr::message::MessageBuffer::get is deprecated, use take")]]
   std::vector<uint8_t> get();
+
+  /**
+   * @brief Moves the whole buffer, leaving MessageBuffer empty.
+   * @returns The buffer as an rvalue-ref.
+   */
   std::vector<uint8_t>&& take();
 
   std::string to_hex() const;
 
   MessageBuffer& operator=(const MessageBuffer& other) = default;
   MessageBuffer& operator=(MessageBuffer&& other) = default;
 
-  template<typename Uint_t>
-  friend MessageBuffer& operator<<(MessageBuffer& msg, Uint_t val);
-  template<typename Uint_t>
-  friend MessageBuffer& operator>>(MessageBuffer& msg, Uint_t& val);
+  /**
+   * @brief A fancy operator for push, writes a byte on the end of the buffer.
+   * @param buffer The MessageBuffer to push to.
+   * @param value The byte to push.
+   * @returns The MessageBuffer that was written to.
+   */
+  friend inline MessageBuffer& operator<<(MessageBuffer& buffer, uint8_t value)
+  {
+    buffer._buffer.push_back(value);
+    return buffer;
+  }
 
-private:
-  std::vector<uint8_t> _buffer;
-};
+  /**
+   * @brief Writes QUICR integral types to the buffer in NBO.
+   * @tparam T A type satisfying quicr::is_integral.
+   * @param msg The MessageBuffer to write to.
+   * @param value The message to be written.
+   * @returns The MessageBuffer that was written to.
+   */
+  template<typename T, typename = std::enable_if_t<quicr::is_integral_v<T>, T>>
+  friend inline MessageBuffer& operator<<(MessageBuffer& msg, T value)
+  {
+    value = swap_bytes(value);
+    uint8_t* val_ptr = reinterpret_cast<uint8_t*>(&value);
 
-MessageBuffer&
-operator<<(MessageBuffer& msg, const uintVar_t& val);
-MessageBuffer&
-operator>>(MessageBuffer& msg, uintVar_t& val);
+    const auto length = msg._buffer.size();
+    msg._buffer.resize(length + sizeof(T));
+    std::memcpy(msg._buffer.data() + length, val_ptr, sizeof(T));
 
-MessageBuffer&
-operator<<(MessageBuffer& msg, const std::vector<uint8_t>& val);
-MessageBuffer&
-operator<<(MessageBuffer& msg, std::vector<uint8_t>&& val);
-MessageBuffer&
-operator>>(MessageBuffer& msg, std::vector<uint8_t>& val);
+    return msg;
+  }
 
-namespace {
-// clang-format off
-constexpr uint16_t
-swap_bytes(uint16_t value)
-{
-  if constexpr (std::endian::native == std::endian::big)
-    return value;
+  /**
+   * @brief Overload for Namespace integral type to write in NBO.
+   * @param msg The MessageBuffer to write to.
+   * @param value The message to be written.
+   * @returns The MessageBuffer that was written to.
+   */
+  friend MessageBuffer& operator<<(MessageBuffer& msg, quicr::Namespace value)
+  {
+    if constexpr (std::endian::native == std::endian::big)
+      return msg << value.name() << value.length();
```

Comment:
These should be written the same regardless of architecture. The actual values might have a different byte order.